src/main/java/org/springframework/integration/kafka/inbound/KafkaMessageDrivenEndpoint.java

```diff
@@ -271,6 +303,19 @@ public int afterShutdown() {
 		return getPhase();
 	}
 
+	@SuppressWarnings("unchecked")
+	private void exchange(Message<?> message) {
+		Object replyObject = sendAndReceive(message);
+		V reply = null;
+		try {
+			reply = (V) replyObject;
+		}
+		catch (ClassCastException e) {
+			logger.error("Unexpected reply type: " + replyObject.getClass().getName(), e);
+		}
+		this.replyTemplate.send(getMessageBuilderFactory().withPayload(reply).build());
```

Comment:
I understand that's not the purpose of the key; it works, but it's not clear how it hurts if every key is different.

In any case, I think the approach we talked about - embedding headers in some way - is probably a better way to go. I will rework some of this while you are on PTO.